utils/terrascript_client.py

```diff
@@ -1726,3 +1735,183 @@ def unpack_namespace_info(self, namespace_info):
         cluster = namespace_info['cluster']['name']
         namespace = namespace_info['name']
         return cluster, namespace
+
+    @staticmethod
+    def validate_elasticsearch_version(version):
+        """ Validate ElasticSearch version. """
+        return version in [7.4, 7.1,
+                           6.8, 6.7, 6.5, 6.4, 6.3, 6.2, 6.0,
+                           5.6, 5.5, 5.3, 5.1,
+                           2.3,
+                           1.5]
+
+    @staticmethod
+    def get_elasticsearch_service_role_tf_resource(identifier):
+        """ Service role for ElasticSearch. """
+        service_role = {
+          'aws_service_name': "es.amazonaws.com"
+        }
+        return aws_iam_service_linked_role(identifier, **service_role)
+
+    @staticmethod
+    def is_elasticsearch_domain_name_valid(name):
+        """ Handle for Error creating Elasticsearch:
+        InvalidParameterValue: Elasticsearch domain name must must start with a
+        lowercase letter and must be between 3 and 28 characters. Valid
+        characters are a-z (lowercase only), 0-9, and - (hyphen). """
+        if len(name) < 3 or len(name) > 28:
+            return False
+        pattern = r'^[a-z][a-z0-9-]+$'
+        return re.search(pattern, name)
+
+    def populate_tf_resource_elasticsearch(self, resource, namespace_info,
+                                           existing_secrets):
+
+        account, identifier, values, output_prefix, output_resource_name = \
+            self.init_values(resource, namespace_info)
+
+        tf_resources = []
+
+        self.init_common_outputs(tf_resources, namespace_info,
+                                 output_prefix, output_resource_name)
+
+        if values['identifier'] and \
+           not self.is_elasticsearch_domain_name_valid(values['identifier']):
+            raise ElasticSearchResourceNameInvalidError(
+                f"[{account}] ElasticSearch domain name must must start with" +
+                f" a lowercase letter and must be between 3 and 28 " +
+                f"characters. Valid characters are a-z (lowercase only), 0-9" +
+                f", and - (hyphen). " +
+                f"{values['identifier']}")
+
+        # values['domain_name'] = identifier
+        elasticsearch_version = values.get('elasticsearch_version')
```

Comment:
I would provide a default or this will throw an exception if elasticsearch_version isn't defined

Reply:
I had defined a default but removed it. Without the error, user may assume their ES cluster is setup with the version they provided.

Reply:
This effectively makes elasticsearch_version a required config param, but it is in the defaults yaml.  I would recommend moving version to the ns yaml level if we want it to be required.  Most required config params don't live in the defaults yaml.

Reply:
Yep that makes sense. I think I will default to latest `7.4` if none is provided. However, a value is provided, I will run the check and make sure it is a valid version.

Reply:
i would prefer to keep the namespace definition as clean as possible, and add schema validation to the defaults file. does that make sense? we should do that to many other resources as well.

Reply:
@maorfr Sounds good. I will refactor the code later once I sync with you on this.